### ISSUE TITLE
Master defence against advanced attackers?

### ISSUE THREAD

**Issue by Zorlin** [lizardfs/lizardfs#416]
Hey there,

I'm currently laying the foundation for some wargames to possibly be run inside GetGlass and other organizations. Basically a testing exercise to ensure our admins know how to deal with various disasters.

Two of the situations I would like to test is that of an attacker intentionally deleting the metadata set on 1) just the master or 2) the master and all shadows.

The valid response to situation 1 would be to fail over to a working shadow that still has the metadata set, and the response to situation 2 would be to recover the metadata set from the metaloggers and use it to restore the filesystem.

My question is what happens in such a scenario, with regard to the empty masters... Would there be a risk of chunkservers deleting chunks, thinking they are no longer needed? And if so, how do we protect against such a scenario?

Perhaps chunkservers could (by default) track the metadata version and refuse to delete chunks while the metadata version is lower than they "remember". This could of course be tuned.

Let me know your thoughts. Thanks.

**Comment by borkd** [comment]
+1 for any tunable failsafe measures in general. That would be a useful first line of defense against hardware or software glitches, or an accidental operations issue (automated or not).

It is my understanding that as of today all chunkservers will follow through on information obtained from a single master, identified by site specific address and port number. A number of scenarios comes to mind where nefarious and determined admin on real, or made up master server could lead to chunks being dropped by chunkservers. If such scenario played out all consumers of particular lizardfs storage cluster instance would be affected and subject to unplanned downtime.

PITA with its own challenges, especially on busy storage, but could help reduce the length of downtime caused by malicious metadata manipulation:
 - chunks: Keep a rolling set of immutable snapshots on filesystems used by chunkservers, monitor space utilization. Tune expiration schedule and triggers to match your site's needs, i.e.: only after "all clear" response from a set of canaries utilizing lizardfs filesystem you are trying to protect.
 - metadata: Ensure you have at least one operational metalogger with plenty of available free space.

One can hope that metadata service will eventually become truly distributed and with proper deployment less likely to be altered everywhere the same time. Restarts with altered metadata (malicious or not) would be taken care of during voting by consensus and before becoming the Source of Truth.
